File: 프로그래머스 고득점 kit/그래프/가장 먼 노드.py
# 완탐 문제
from collections import deque, Counter


def solution(n, vertex):
    g = [[] for _ in range(n + 1)]
    for v, e in vertex:
        g[v].append(e)
        g[e].append(v)

    q = deque([1])
    visited = [0] * (n + 1)
    visited[1] = 1

    while q:
        curr = q.popleft()

        for next in g[curr]:
            if not visited[next]:
                visited[next] = visited[curr] + 1
                q.append(next)

    m = max(visited)
    return visited.count(m)


print(solution(6, [[3, 6], [4, 3], [3, 2], [
      1, 3], [1, 2], [2, 4], [5, 2]]) == 3)

File: 프로그래머스 고득점 kit/그래프/가장 먼 노드.js
/** @param {number} n @param {[number,number][]} edge*/
function solution(n, edge) {
  const g = Array.from({ length: n + 1 }, () => new Set([]))
  const visited = Array.from({ length: n + 1 }, () => 0)
  visited[1] = 1

  edge.forEach(([v, e]) => {
    g[v].add(e)
    g[e].add(v)
  })

  const q = [1]

  while (q.length > 0) {
    const curr = q.shift()
    for (const next of g[curr]) {
      if (!visited[next]) {
        visited[next] = visited[curr] + 1
        q.unshift(next)
      }
    }
  }

  const maximum = Math.max(...visited)
  return visited.filter((e) => e === maximum).length
}

console.log(
  solution(6, [
    [3, 6],
    [4, 3],
    [3, 2],
    [1, 3],
    [1, 2],
    [2, 4],
    [5, 2],
  ]) === 3
)
